lib/Serialization/ModuleFileSharedCore.cpp

```diff
@@ -334,6 +334,12 @@ static ValidationInfo validateControlBlock(
         LLVM_FALLTHROUGH;
       case 3:
         result.shortVersion = blobData.slice(0, scratch[2]);
+
+        // If the format version doesn't match, give up after also getting the
+        // compiler version. This provides better diagnostics.
+        if (result.status != Status::Valid)
+          return result;
+
```

Comment:
As I said in the other PR, not sure about how any of this work, I just want to understand the consequences of this change. Wouldn't this make working with a compiler from `main` impossible to use against files compiled with previous compilers? That would make using `.swiftmodule` files distributed by Xcode not valid, wouldn't it? Before this I think old files were still mostly compatible.

Reply:
This won't change the current behavior. I would not expect a main compiler to be able to use swiftmodules built by any compiler older than a week or so. If that swiftmodule has a different serialization format it would have been rejected already.

The main content of swiftmodule files have very limited compatibility as they are designed to be a local cache. In theory they should only be read by the exact same compiler that wrote them, and in the exact same context. That's why any distributed modules should use swiftinterfaces. And the only swiftmodules shipped with Xcode are only the prebuilt cache, they can all be rebuilt from swiftinterfaces in a similar manner by newer compilers.

So any swiftmodule with a different serialization format version are rejected. swiftmodule files are also rejected between released compilers with a different tag. We leave some room for dev compilers, just to avoid rebuilding the stdlib all the time during development.

As you pointed out, the control block should be compatible with different compilers. But as soon as we know it should be rejected it's safe to skip reading the rest of the information.